File: pulsar-skeleton/src/main/java/ai/platon/pulsar/common/GlobalExecutor.java
package ai.platon.pulsar.common;

import ai.platon.pulsar.common.config.CapabilityTypes;
import ai.platon.pulsar.common.config.ImmutableConfig;
import ai.platon.pulsar.common.config.Params;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicBoolean;

public class GlobalExecutor implements AutoCloseable {

    public static final Logger LOG = LoggerFactory.getLogger(GlobalExecutor.class);
    /**
     * The number of processors available to the Java virtual machine
     */
    public static final int NCPU = Runtime.getRuntime().availableProcessors();
    private static GlobalExecutor INSTANCE;
    private ExecutorService executor;
    private AtomicBoolean closed = new AtomicBoolean(false);
    private float autoConcurrencyFactor;
    private int concurrency;

    private GlobalExecutor(ImmutableConfig immutableConfig) {
        int concurrencyHint = immutableConfig.getInt(CapabilityTypes.GLOBAL_EXECUTOR_CONCURRENCY_HINT, -1);
        concurrency = concurrencyHint;
        if (concurrency <= 0) {
            autoConcurrencyFactor = immutableConfig.getFloat(CapabilityTypes.GLOBAL_EXECUTOR_AUTO_CONCURRENCY_FACTOR, 1);
            concurrency = Math.max((int) (NCPU * autoConcurrencyFactor), 4);
        }

        Params.of(
                "availableProcessors", NCPU,
                "autoConcurrencyFactor", autoConcurrencyFactor,
                "concurrencyHint", concurrencyHint,
                "concurrency", concurrency
        ).withLogger(LOG).info(true);
    }

    public static GlobalExecutor getInstance(ImmutableConfig immutableConfig) {
        if (INSTANCE == null) {
            INSTANCE = new GlobalExecutor(immutableConfig);
        }

        return INSTANCE;
    }

    public int getConcurrency() {
        return concurrency;
    }

    /**
     * TODO: Allocate executors for sessions separately
     * TODO: Check
     */
    public ExecutorService getExecutor() {
        if (executor == null) {
            initExecutorService();
        }

        return executor;
    }

    private void initExecutorService() {
        synchronized (GlobalExecutor.class) {
            if (executor == null) {
                executor = Executors.newWorkStealingPool(concurrency);
            }
        }
    }

    @Override
    public void close() {
        if (closed.getAndSet(true)) {
            return;
        }

        if (executor != null && !executor.isShutdown()) {
            executor.shutdownNow();
        }
    }
}

File: pulsar-skeleton/src/main/kotlin/ai/platon/pulsar/common/GlobalExecutor.kt
package ai.platon.pulsar.common

import ai.platon.pulsar.common.config.AppConstants
import ai.platon.pulsar.common.config.CapabilityTypes
import ai.platon.pulsar.common.config.ImmutableConfig
import ai.platon.pulsar.common.config.Params
import org.slf4j.LoggerFactory
import java.util.concurrent.ExecutorService
import java.util.concurrent.ForkJoinPool
import java.util.concurrent.ForkJoinPool.ForkJoinWorkerThreadFactory
import java.util.concurrent.Future
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.math.max

class GlobalExecutor(conf: ImmutableConfig) : AutoCloseable {
    val log = LoggerFactory.getLogger(GlobalExecutor::class.java)

    private var executor: ExecutorService? = null
    private val closed = AtomicBoolean(false)
    private var autoConcurrencyFactor = conf.getFloat(CapabilityTypes.GLOBAL_EXECUTOR_AUTO_CONCURRENCY_FACTOR, 1f)
    private val threadFactory = ForkJoinWorkerThreadFactory { pool ->
        ForkJoinPool.defaultForkJoinWorkerThreadFactory.newThread(pool).also { it.name = String.format("w%02d", it.poolIndex) }
    }
    private val concurrencyHint = conf.getInt(CapabilityTypes.GLOBAL_EXECUTOR_CONCURRENCY_HINT, -1)
    val concurrency = if (concurrencyHint > 0) concurrencyHint else max((AppConstants.NCPU * autoConcurrencyFactor).toInt(), 4)

    init {
        Params.of(
                "availableProcessors", AppConstants.NCPU,
                "autoConcurrencyFactor", autoConcurrencyFactor,
                "concurrencyHint", concurrencyHint,
                "concurrency", concurrency
        ).withLogger(log).info(true)
    }

    fun getExecutor(): ExecutorService {
        synchronized(GlobalExecutor::class.java) {
            if (executor == null) {
                executor = newWorkStealingPool(concurrency)
            }
            return executor!!
        }
    }

    fun <T> submit(task: () -> T): Future<T> {
        return getExecutor().submit(task)
    }

    /**
     * copy from [java.util.concurrent.Executors]
     * */
    private fun newWorkStealingPool(parallelism: Int): ExecutorService {
        return ForkJoinPool(parallelism, threadFactory, null as Thread.UncaughtExceptionHandler?, true)
    }

    override fun close() {
        if (closed.compareAndSet(false, true)) {
            executor?.shutdown()
        }
    }
}
